[Chen Anqi] Duke Increments

anqichen9856 opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

j-lum committed:
Add configuration for console applications

Gradle defaults to an empty stdin which results in runtime exceptions
when attempting to read from `System.in`. Let's add some sensible
defaults for students who may still need to work with the standard
input stream.

j-lum committed:
Merge pull request #13 from j-lum/javaexec

Add configuration for console applications

Commit:
Add greet, echo and exit functions. Indentation and horizontal lines omitted.

Commit:
Add the ability to store text entered by the user and display them in order when the user types "list".

Commit:
Add Task class to facilitate marking tasks as done.

Commit:
Add the ability to mark tasks as done.

Commit:
Use Inheritance to support multiple task types (i.e. implement classes Todo, Deadline and Event classes to inherit from a Task class).

Commit:
Add support for tracking three types of tasks:
1. ToDos: tasks without any date/time attached to it e.g., visit new theme park
2. Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
3. Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

Commit:
Use the input/output redirection technique to semi-automate testing of Duke.

Commit:
Define a class DukeException to represent exceptions specific to Duke.

Commit:
Use DukeException to handle errors such as incorrect inputs entered by the user. Errors handled include: incorrect commands (other than "todo" "event" "deadline" "done" "list" "bye"); missing description for ToDos, Deadlines and Events; missing time for Deadlines and Events; missing task number for Done.

Commit:
Add support for deleting tasks from the list. Update error handling to include missing task number for "delete".

Commit:
Add the function to save the tasks in the hard disk automatically whenever the task list changes. Every time when Duke starts up, load the data from the hard disk.

Commit:
Add the ability to parse dates and times of the format "dd/MM/yyyy HHmm" (e.g. 02/12/2019 1800) as a LocalDateTime object instead of a simple String.

anqichen9856 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add the function to save the tasks in the hard disk automatically whenever the task list changes. Every time when Duke starts up, load the data from the hard disk.

anqichen9856 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add the ability to parse dates and times of the format "dd/MM/yyyy HHmm" (e.g. 02/12/2019 1800) as a LocalDateTime object instead of a simple String.

# Conflicts:
#	src/main/java/Deadline.class
#	src/main/java/Event.class

anqichen9856 committed:
Make the code more OOP by extracting out closely related code as classes.
Extracted the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list
*Command classes (i.e., AddCommand, DeleteCommand, ExitCommand etc.) that inherits from an abstract Command class

anqichen9856 committed:
Organize the classes into suitable java packages. e.g., duke.task, duke.command, duke.core

anqichen9856 committed:
Add JUnit tests to test the behavior of the code.

anqichen9856 committed:
Add JavaDoc comments to the code.

anqichen9856 committed:
Tweak the code to comply with a chosen coding standard.

anqichen9856 committed:
Add the ability to find a task by searching for a keyword.

anqichen9856 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments to the code.

anqichen9856 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a chosen coding standard.

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/core/DukeException.java
#	src/main/java/duke/core/Parser.java
#	src/main/java/duke/core/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/ToDo.java

anqichen9856 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add the ability to find a task by searching for a keyword.

# Conflicts:
#	src/main/java/duke/core/Parser.class
#	src/main/java/duke/core/Parser.java
#	src/main/java/duke/core/Ui.class
#	src/main/java/duke/core/Ui.java

anqichen9856 committed:
Add Javadoc comment to Level-9 classes and methods. Add an empty line between description and parameter section for every Javadoc comment.

anqichen9856 committed:
Merge branches 'master' and 'gradle' of https://github.com/anqichen9856/duke

* 'master' of https://github.com/anqichen9856/duke:

* 'gradle' of https://github.com/anqichen9856/duke:
  Add configuration for console applications
  Change file mode on `gradle` to be executable
  Add sample checkstyle configuration
  Add support for Gradle workflow

seeziyang commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+     * @return A boolean. True if the command tells Duke to exit, false 
+     *          otherwise.      
+     */
+    public abstract boolean isExit();
```
You might want to consider having an implementation for this method by having it return false and only overriding in the ExitCommand class to return true.

seeziyang commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,131 @@
+     */
+    private static void checkIllegalInstruction(String[] words) throws DukeException {
+        String fw = words[0];
+        if (!(fw.equals("done") || fw.equals("todo") || fw.equals("deadline") || fw.equals("event")
```
Having the default case of your switch statement below throw the illegal instruction exception might be a better option instead of manually checking if the first word is not any valid command.

seeziyang commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+        try {
+            dateTime = LocalDateTime.parse(by, f);
+        } catch (DateTimeParseException e) {
+        }
```
Might not be a good idea to have an empty catch block, perhaps you can print out the error?

seeziyang reviewed (commented):
Great work!! 👍

Code was clear and easily understandable. I did not see any coding style violation and there was ample javadocs.

Minor suggestion:
- Organize your import statements to keep the ordering consistent and group them based on package imported from as per [https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html#package-and-import-statements](https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html#package-and-import-statements)

anqichen9856 committed:
Automate project builds using Gradle

Set up gradle to build and run Duke. Set up gradle to run unit tests. Add Shadow plugin plugin that packages an application into an executable jar file. Automate the management of dependencies to third-party libraries such as JUnit.

anqichen9856 committed:
Use checkStyle to detect coding style violations.

Add checkstyle plugin to gradle to check if the code complies with a set of style rules specified in the configuration files.

anqichen9856 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Use checkStyle to detect coding style violations.

b123b123z reviewed (commented):
Good job! 
Code was tweaked according to chosen coding standard and checkstyle plugin was also successfully implemented with gradle to ensure proper styling of code which made it easier for me to read and understand in addition to helpful javadocs comments. 

Overall, the code is robust and easy to understand. I did not spot any major errors!

anqichen9856 committed:
Add a GUI to Duke

Use JavaFX to create a GUI for Duke (without using FXML).

anqichen9856 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add a GUI to Duke

anqichen9856 commented:
> Great work!! 👍
> 
> Code was clear and easily understandable. I did not see any coding style violation and there was ample javadocs.
> 
> Minor suggestion:
> 
> * Organize your import statements to keep the ordering consistent and group them based on package imported from as per https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html#package-and-import-statements

Thank you for your comments! They're really helpful. I will modify my code according to your review.

anqichen9856 commented:
> Good job!
> Code was tweaked according to chosen coding standard and checkstyle plugin was also successfully implemented with gradle to ensure proper styling of code which made it easier for me to read and understand in addition to helpful javadocs comments.
> 
> Overall, the code is robust and easy to understand. I did not spot any major errors!

Thank you for your input!

b123b123z commented:
Good job!
Code was tweaked according to chosen coding standard and checkstyle plugin was also successfully implemented with gradle to ensure proper styling of code which made it easier for me to read and understand in addition to helpful javadocs comments.

Overall, the code is robust and easy to understand. I did not spot any major errors!

anqichen9856 committed:
Add specific response messages in GUI

Modify the stub getResponse(String input) method in Duke so that it returns detailed response instead of "Duke heard...".

anqichen9856 committed:
Use FXML and improve Duke GUI

Use FXML to separate logic and remove dependency between GUI classes. Let Duke show welcome message and print loadSuccess/loadError messages upon activation. Further customize User and Duke dialog boxes (different colors, paddings).

anqichen9856 committed:
Add Assertions (#4)

Use assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code.

anqichen9856 committed:
Improve code quality

Refactor the code to improve the code quality where necessary. Modify complicated boolean expressions in Parser class.

anqichen9856 committed:
Fix regressions caused by refactoring

Fix logic bugs in checkIllegalInstruction method in Parser class. Handle one more exception (task index given in "done" or "delete" commands is out of bounds) from user input.

anqichen9856 committed:
Use Lambdas in Ui class

Use the Lambdas feature of Java for printing all tasks from a TaskList in Ui class.

anqichen9856 committed:
Use Streams in Parser and Ui classes

Use the Streams feature of Java for displaying search results in Ui class, as well as subString and findIdx methods in Parser class.

anqichen9856 committed:
Teach Duke GUI to automatically quit when user types "bye"

Add delayed stage.close() when GUI detects a "bye" input.

anqichen9856 committed:
Teach Duke to reformat user input date and time

Remove the unused dateTime field in Deadline and Event classes. Add the ability to parse user input date and time to a more readable string representation.

anqichen9856 committed:
Add the ability to update existing tasks

Support a way to easily edit details of an existing task without having to delete it first, e.g. change the time of an event without changing anything else. Improve organisation of import statements in various classes.

DivineDX commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,241 @@
+        }
+    }
+
+    private static String timeToAmPm(LocalDateTime dateTime) {
```
Can consider using the SimpleDateFormat class instead of re-inventing the wheel.

DivineDX commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,241 @@
+     */
+    private static void checkIllegalInstruction(String[] words) throws DukeException {
+        String fw = words[0];
+        boolean isDoneCommand = fw.equals("done");
```
You could just use a massive switch statement to check for the first word, and have the default case return an "UnknownCommand" class that tells the user that his/her command is invalid

DivineDX reviewed (commented):
[PR Review Done in Tutorial]

anqichen9856 committed:
Improve code based on peer review

Use DateTimeFormatter to format the output string of date and time instead of manually coding the string. Change the 'by' and 'at' fields in Deadline and Event classes to more descriptive 'dueTime' and 'occurTime' fields.

anqichen9856 committed:
Improve code quality (more understandable naming & shorter methods)

Update class, method and parameter names to follow coding standard more closely:
(1) ToDo (class) -> NormalTask
(2) parseDateTime (method) -> formatDateTime (method)
(3) String s (parameter) -> String instruction / String errorMessage

Make Parser class more concise.

anqichen9856 committed:
Further shorten Parser method

anqichen9856 committed:
Write Javadoc comments for recently added methods and classes

anqichen9856 committed:
Fix Parser bug caused by case fallthrough

anqichen9856 committed:
Set theme jekyll-theme-cayman

anqichen9856 committed:
Set theme jekyll-theme-cayman

anqichen9856 committed:
Add a user guide

Update docs\README.md to introduce Duke features and usage with examples.

anqichen9856 committed:
Merge branch 'master' of https://github.com/anqichen9856/duke

* 'master' of https://github.com/anqichen9856/duke:
  Set theme jekyll-theme-cayman
  Set theme jekyll-theme-cayman

anqichen9856 committed:
Improve user guide formatting

anqichen9856 committed:
Improve user guide formatting

anqichen9856 committed:
Improve user guide formatting

anqichen9856 committed:
Improve user guide formatting

anqichen9856 committed:
Set theme jekyll-theme-cayman

anqichen9856 committed:
Change Ui.png filename

anqichen9856 committed:
Revert "Change Ui.png filename"

This reverts commit b6f4cec0b3080a0c3c30fe2640761a7cbde2d470.

anqichen9856 committed:
Revert "Revert "Change Ui.png filename""

This reverts commit 500a52be6607c93fcb629bb4759c37c384ef2dde.

anqichen9856 committed:
Change Ui.png filename

anqichen9856 committed:
Fix the problem of long dialog being truncated in GUI

Change Label to VBox and Text.

anqichen9856 committed:
Update Ui.png image

anqichen9856 committed:
Update Ui.png to a more representative one

anqichen9856 committed:
Change Duke to Tsinghua

anqichen9856 committed:
Remove magic number in DialogBox class

Replace magic number in DialogBox class with private static final variables. Give credit for stage close delay code in Main class.

anqichen9856 committed:
Refactor "load" method in Storage class & change local file access method

Shorten the "load" method by moving all code related to parsing to Parser class. Change local file access method to "getResourceAsStream" so that Jar file can find the file.

anqichen9856 committed:
Let the user specify tasks file path upon launching Duke

anqichen9856 committed:
Add missing Javadoc and fix checkstyle errors

anqichen9856 committed:
Clear messy logic surrounding file I/O, and remove dead code from stages before GUI

anqichen9856 committed:
Update user guide to match the latest Duke functions

anqichen9856 committed:
Minor modifications to GUI appearance

anqichen9856 committed:
Change Duke personality

anqichen9856 committed:
Update Ui.png

anqichen9856 committed:
Add multi-platform support

anqichen9856 committed:
Teach Duke to parse and format both start time & end time of an event

User input for event now include both start time and end time, instead of just a start time.

anqichen9856 committed:
Fix regression in "update" command caused by new parsing method for event time

anqichen9856 committed:
Update User Guide and Ui.png to reflect latest changes in input and output formats

anqichen9856 committed:
Improve User Guide

anqichen9856 committed:
Update User Guide

anqichen9856 committed:
Update User Guide
